sql: add KV stats to TraceAnalyzer

cathymw opened the pull request:
The first commit adds total bytes and total rows read from KV to the TraceAnalyzer.
Closes: #56605

The second commit adds cumulative time spent in KV to the TraceAnalyzer.
Closes: #56604

Release note: None.

cathymw committed:
sql: add rows/bytes read from KV to TraceAnalyzer

This commit adds total bytes and total rows read from KV to the
TraceAnalyzer.

Release note: None.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/57495)
<!-- Reviewable:end -->

cathymw committed:
sql: add cumulative KV time to TraceAnalyzer

This commit adds the total time spent waiting for KV requests to
TraceAnalyzer.

Release note: None.

cathymw reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57495)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/execinfrapb/component_stats.go, line 149 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57495#-MNeg_5wCsANOw9yeF8-:-MNeg_5wCsANOw9yeF80:b90rjij) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ad5c711fccb141117d816ca0ae0b7e44577b8ab/pkg/sql/execinfrapb/component_stats.go#L149)):*
> ```Go
> 	timeVal := func(v *optional.Duration) {
> 		if v.HasValue() {
> 			v.Set(1)
> ```

While I was adding a test to `traceanalyzer_test.go` for time spent in KV, I was thinking that setting time stat values to a non-zero value in `MakeDeterministic()` would maybe make it more useable for testing, since it would differentiate components that have stats that may be non-deterministic vs. components that never set time stats. This would probably require a bunch of tests to be updated. Not sure if there's anything else I'm not considering or if there's a better way to test time valued stats. Thoughts?


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
___
*[pkg/sql/execinfrapb/component_stats.go, line 149 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/57495#-MNeg_5wCsANOw9yeF8-:-MNensNFEMy7Q53tuWoU:b-71gnim) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ad5c711fccb141117d816ca0ae0b7e44577b8ab/pkg/sql/execinfrapb/component_stats.go#L149)):*
<details><summary><i>Previously, cathymw wrote…</i></summary><blockquote>

While I was adding a test to `traceanalyzer_test.go` for time spent in KV, I was thinking that setting time stat values to a non-zero value in `MakeDeterministic()` would maybe make it more useable for testing, since it would differentiate components that have stats that may be non-deterministic vs. components that never set time stats. This would probably require a bunch of tests to be updated. Not sure if there's anything else I'm not considering or if there's a better way to test time valued stats. Thoughts?
</blockquote></details>

I think a better way to test the trace analyzer is to add infrastructure to be able to specify `ComponentStats` in the test and associate them with the components in the test flow.


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

The TraceAnalyzer refactoring looks great. One thing to note is that we're duplicating a lot of boilerplate for calculating each stat, maybe we want to just have a single "Process" function that calculates all the stats and stores them in the TraceAnalyzer (and which can then be accessed as needed).

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57495#-:-MNyvxqp3alCtskZKzeB:b-a6vmp6)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @asubiotto, @cathymw, @RaduBerinde, and @yuzefovich)



<!-- Sent from Reviewable.io -->

asubiotto commented:
Waiting to review this until #57711 is in, since this will probably need a rebase on top of that.

cathymw commented:
RFAL! :)

asubiotto reviewed (changes requested):
> Reviewed 14 of 14 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57495)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @asubiotto, @cathymw, and @yuzefovich)

___
*[pkg/sql/execstats/traceanalyzer.go, line 26 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/57495#-MOlJt5F-z-DiYbRsiHw:-MOlJt5F-z-DiYbRsiHx:b2aiqp) ([raw file](https://github.com/cockroachdb/cockroach/blob/e3eb9bdedba091a01e787d88aab3b1597feb192a/pkg/sql/execstats/traceanalyzer.go#L26)):*
> ```Go
> 
> // ProcessorStats contains stats for a specific processor extracted from a trace.
> type ProcessorStats struct {
> ```

I don't think there's a need to make these structs public. What about adding a function to a testing file in this package as described in the following comment that adds stats with an accompanying node id. Something like:
in `utils_test.go`:
```
func (a *TraceAnalyzer) AddComponentStats(nodeID roachpb.NodeID, stats *execinfrapb.ComponentStats) {
    <modify TraceAnalyzer internal state to add stats for the processor/stream/flow specified in stats.ComponentID and the given node ID>
}
```

This would also allow us to assert node-level stats later on.
___
*[pkg/sql/execstats/traceanalyzer.go, line 44 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/57495#-MOkuNsS-Fh4-7PwMENN:-MOkuNsS-Fh4-7PwMENO:bcmdlfe) ([raw file](https://github.com/cockroachdb/cockroach/blob/e3eb9bdedba091a01e787d88aab3b1597feb192a/pkg/sql/execstats/traceanalyzer.go#L44)):*
> ```Go
> // SetProcessorStats creates a ProcessorStats with the given processor information. This is used
> // for testing.
> func SetProcessorStats(nodeID roachpb.NodeID, stats *execinfrapb.ComponentStats) *ProcessorStats {
> ```

These package-level functions shouldn't be added to the main package if they're only going to be used by testing code. There's a neat trick where you can create a new `utils_test.go` file (or any file with a `_test.go` suffix and add a function like this there. This will make it so that testing code from other packages (i.e. `execstats_test`) can call these functions, but not non-test code.
___
*[pkg/sql/execstats/traceanalyzer_test.go, line 178 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/57495#-MOksMHK62ib0lSte9kR:-MOksMHL7lMECrV9l0cX:b-8ot5bh) ([raw file](https://github.com/cockroachdb/cockroach/blob/c1450a78a423cfd41b93b88b167cf4601e837eb5/pkg/sql/execstats/traceanalyzer_test.go#L178)):*
> ```Go
> 			rowexecTraceAnalyzer, colexecTraceAnalyzer,
> 		} {
> 			queryLevelStats := analyzer.GetQueryLevelStats()
> ```

nit: it feels like this subtest approach is not going to scale well. I think it's probably best to just remove these subtests, get the querylevelstats once above and then assert all the values one the same stats object rather than creating a new subtest for each stat.


<!-- Sent from Reviewable.io -->

cathymw reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57495)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @asubiotto and @yuzefovich)

___
*[pkg/sql/execstats/traceanalyzer.go, line 26 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/57495#-MOlJt5F-z-DiYbRsiHw:-MOmZEGJBfzBzWcLWj4Y:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e3eb9bdedba091a01e787d88aab3b1597feb192a/pkg/sql/execstats/traceanalyzer.go#L26)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

I don't think there's a need to make these structs public. What about adding a function to a testing file in this package as described in the following comment that adds stats with an accompanying node id. Something like:
in `utils_test.go`:
```
func (a *TraceAnalyzer) AddComponentStats(nodeID roachpb.NodeID, stats *execinfrapb.ComponentStats) {
    <modify TraceAnalyzer internal state to add stats for the processor/stream/flow specified in stats.ComponentID and the given node ID>
}
```

This would also allow us to assert node-level stats later on.
</blockquote></details>

Done.
___
*[pkg/sql/execstats/traceanalyzer.go, line 44 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/57495#-MOkuNsS-Fh4-7PwMENN:-MOmZB4Q8zHx6zAw6by1:b-pyz1he) ([raw file](https://github.com/cockroachdb/cockroach/blob/e3eb9bdedba091a01e787d88aab3b1597feb192a/pkg/sql/execstats/traceanalyzer.go#L44)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

These package-level functions shouldn't be added to the main package if they're only going to be used by testing code. There's a neat trick where you can create a new `utils_test.go` file (or any file with a `_test.go` suffix and add a function like this there. This will make it so that testing code from other packages (i.e. `execstats_test`) can call these functions, but not non-test code.
</blockquote></details>

Done. TIL
___
*[pkg/sql/execstats/traceanalyzer_test.go, line 178 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/57495#-MOksMHK62ib0lSte9kR:-MOmZ8NfDUjx6sZH_9sB:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c1450a78a423cfd41b93b88b167cf4601e837eb5/pkg/sql/execstats/traceanalyzer_test.go#L178)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

nit: it feels like this subtest approach is not going to scale well. I think it's probably best to just remove these subtests, get the querylevelstats once above and then assert all the values one the same stats object rather than creating a new subtest for each stat.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

asubiotto reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> once comments are addressed

> Reviewed 3 of 3 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57495#-:-MOpzzOJC7IQq2WaANPE:bdqxoaz)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @asubiotto, @cathymw, and @yuzefovich)

___
*[pkg/sql/execstats/traceanalyzer_test.go, line 175 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/57495#-MOpxB_Q-S1g2pGmNMUq:-MOpxB_Q-S1g2pGmNMUr:bjprmkr) ([raw file](https://github.com/cockroachdb/cockroach/blob/4bccd5695f520d4c2c5566509209101e54d5fa68/pkg/sql/execstats/traceanalyzer_test.go#L175)):*
> ```Go
> }
> 
> func TestTraceAnalyzer_ProcessStats(t *testing.T) {
> ```

nit: remove the underscore
___
*[pkg/sql/execstats/traceanalyzer_test.go, line 178 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/57495#-MOpwklW8DOXgL03ksMY:-MOpwklW8DOXgL03ksMZ:bht16xz) ([raw file](https://github.com/cockroachdb/cockroach/blob/4bccd5695f520d4c2c5566509209101e54d5fa68/pkg/sql/execstats/traceanalyzer_test.go#L178)):*
> ```Go
> 	componentStats1 := &execinfrapb.ComponentStats{
> 		Component: execinfrapb.ComponentID{
> 			FlowID: execinfrapb.FlowID{UUID: uuid.MakeV4()},
> ```

You have a couple of helpers you can use in `ProcessorComponentID` and the like instead of making a `ComponentID` yourself.
___
*[pkg/sql/execstats/traceanalyzer_test.go, line 203 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/57495#-MOpwwJg0yHGmI93qsAH:-MOpwwJh9WEe0aGE1Puv:b157gju) ([raw file](https://github.com/cockroachdb/cockroach/blob/4bccd5695f520d4c2c5566509209101e54d5fa68/pkg/sql/execstats/traceanalyzer_test.go#L203)):*
> ```Go
> 
> 	a := &execstats.TraceAnalyzer{FlowMetadata: &execstats.FlowMetadata{}}
> 	a.AddComponentStats(1, componentStats1)
> ```

This is more of a style issue, but why not create the component stats directly here? This would avoid having to come up with local variable names with integers. Also make sure to comment any arguments that are not obvious:
```
a.AddComponentStats(1 /* nodeID */, &execinfrapb.ComponentStats{})
```


<!-- Sent from Reviewable.io -->

cathymw commented:
TFTRs!
bors r=asubiotto
